drivers/spi/spi_pl022.c

```diff
@@ -0,0 +1,499 @@
+ */
+
+#include <board.h>
+#include <errno.h>
+#include <gpio.h>
+#include <spi.h>
+
+
+/* Values for Control Register 0, SSPCR0 */
+
+#define SSPCR0_DSS(dss)	((dss) << 0)	/* Data Size Select (num bits - 1) */
+#define SSPCR0_FRF_SPI	(0 << 4)	/* Motorola SPI frame format */
+#define SSPCR0_FRF_TI	(1 << 4)	/* TI synchronous serial frame format */
+#define SSPCR0_FRF_MW	(2 << 4)	/* National Microwire frame format */
+#define SSPCR0_SPO	(1 << 6)	/* SPI clock polarity */
+#define SSPCR0_SPH	(1 << 7)	/* SPI clock phase */
+#define SSPCR0_SCR(scr)	((scr) << 8)	/* Serial clock rate */
+
+/* Values for Control Register 1, SSPCR1 */
+
+#define SSPCR1_LBM	(1 << 0)	/* Loop back mode */
+#define SSPCR1_SSE	(1 << 1)	/* Synchronous serial port enable */
+#define SSPCR1_MS	(1 << 2)	/* Master or slave mode select */
+#define SSPCR1_SOD	(1 << 3)	/* Slave-mode output disable */
+
+/* Values for Status Register, SSPSR */
+
+#define SSPSR_TFE	(1 << 0)	/* Transmit FIFO empty */
+#define SSPSR_TNF	(1 << 1)	/* Transmit FIFO not full */
+#define SSPSR_RNE	(1 << 2)	/* Receive FIFO not empty */
+#define SSPSR_RFF	(1 << 3)	/* Receive FIFO full */
+#define SSPSR_BSY	(1 << 4)	/* PrimeCell SSP busy flag */
+
+/* Values for interrupt mask/status registers SSPIMSC, SSPRIS and SSPMIS */
+
+#define SSP_INT_ROR	(1 << 0)	/* Receive overrun interrupt flag */
+#define SSP_INT_RT	(1 << 1)	/* Receive timeout interrupt flag */
+#define SSP_INT_RX	(1 << 2)	/* Receive FIFO interrupt flag */
+#define SSP_INT_TX	(1 << 3)	/* Transmit FIFO interrupt flag */
+
+/* Values for Interrupt Clear Register, SSPICR */
+
+#define SSPICR_RORIC	SSP_INT_ROR	/* Clear Receive overrun interrupt */
+#define SSPICR_RTIC	SSP_INT_RT	/* Clear Receive timeout interrupt */
+
+/* Values for DMA Control Register, SSPDMACR */
+
+#define SSPDMACR_RXDMAE	(1 << 0)	/* Receive DMA Enable */
+#define SSPDMACR_TXDMAE	(1 << 1)	/* Transmit DMA Enable */
+
+#define FIFO_DEPTH	8		/* Number of entries in data FIFO */
+
+
+/*
+ * Hardware register layout,
+ */
+struct spi_pl022 {
+	volatile u32_t cr0;	/* Control register 0, macros SSPCR0_ */
+	volatile u32_t cr1;	/* Control register 1, macros SSPCR1_ */
+	volatile u32_t dr;	/* Data (Transmit/Receive FIFO) */
+	volatile u32_t sr;	/* Status register, macros SSPCR_  */
+	volatile u32_t cpsr;	/* Clock prescale register */
+	volatile u32_t imsc;	/* Interrupt mask set/clr, macros SSP_INT_ */
+	volatile u32_t ris;	/* Raw interrupt status, macros SSP_INT_ */
+	volatile u32_t mis;	/* Masked interrupt status, macros SSP_INT_ */
+	volatile u32_t icr;	/* Interrupt clear register, macros SSPICR_ */
+	volatile u32_t dmacr;/* DMA control register, macros SSPDMACR_ */
+};
+
+struct spi_pl022_context;
+
+typedef	void (*spi_pl022_pio_fn_t)(struct spi_pl022_context *state,
+				struct spi_pl022 *ssp);
+
+/* Device config */
+struct spi_pl022_config {
+	struct spi_pl022 *ssp;
+	unsigned int clk_freq;
+	void (*irq_config_func)();
+	const char *cs_gpio_name;
+	u32_t cs_gpio_pin;
```

Comment:
It's all about the controller (if is can control by itself CS lines), and how the slave device is wired as well (if it's wired against a GPIO line and not an CS one for instance).

Looking at http://infocenter.arm.com/help/topic/com.arm.doc.ddi0194g/DDI0194G_ssp_pl022_r1p3_trm.pdf it seems that the controller has its own SS line (SSPFSSOUT on master side, SSPFSSIN in slave side), so the gpio based control is not mandatory.